injector/metapodtemplate.go

```diff
@@ -0,0 +1,266 @@
+type MetaContainer struct {
+	Name         string
+	Env          []corev1.EnvVar
+	VolumeMounts []corev1.VolumeMount
+}
+
+// NewMetaPodTemplate coerces the workload object into a MetaPodTemplate following the mapping definition. The
+// resulting MetaPodTemplate may have one or more service bindings applied to it at a time, but should not be reused.
+// The workload must be JSON marshalable.
+func NewMetaPodTemplate(ctx context.Context, workload runtime.Object, mapping *v1alpha3.ClusterWorkloadResourceMappingTemplate) (*MetaPodTemplate, error) {
+	mpt := &MetaPodTemplate{
+		workload: workload,
+		mapping:  mapping,
+
+		Annotations: map[string]string{},
+		Containers:  []MetaContainer{},
+		Volumes:     []corev1.Volume{},
+	}
+
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(workload)
+	if err != nil {
+		return nil, err
+	}
+	uv := reflect.ValueOf(u)
+
+	if err := mpt.getAt(mpt.mapping.Annotations, uv, &mpt.Annotations); err != nil {
+		return nil, err
+	}
+	for i := range mpt.mapping.Containers {
+		cp := jsonpath.New("")
+		if err := cp.Parse(fmt.Sprintf("{%s}", mpt.mapping.Containers[i].Path)); err != nil {
+			return nil, err
+		}
+		cr, err := cp.FindResults(u)
+		if err != nil {
+			// errors are expected if a path is not found
+			continue
+		}
+		for _, cv := range cr[0] {
+			mc := MetaContainer{
+				Name:         "",
+				Env:          []corev1.EnvVar{},
+				VolumeMounts: []corev1.VolumeMount{},
+			}
+
+			if mpt.mapping.Containers[i].Name != "" {
+				// name is optional
+				if err := mpt.getAt(mpt.mapping.Containers[i].Name, cv, &mc.Name); err != nil {
+					return nil, err
+				}
+			}
+			if err := mpt.getAt(mpt.mapping.Containers[i].Env, cv, &mc.Env); err != nil {
+				return nil, err
+			}
+			if err := mpt.getAt(mpt.mapping.Containers[i].VolumeMounts, cv, &mc.VolumeMounts); err != nil {
+				return nil, err
+			}
+
+			mpt.Containers = append(mpt.Containers, mc)
+		}
+	}
+	if err := mpt.getAt(mpt.mapping.Volumes, uv, &mpt.Volumes); err != nil {
+		return nil, err
+	}
+
+	return mpt, nil
+}
+
+// WriteToWorkload applies mutation defined on the MetaPodTemplate since it was created to the workload resource the
+// MetaPodTemplate was created from. This method should generally be called once per instance.
+func (mpt *MetaPodTemplate) WriteToWorkload(ctx context.Context) error {
+	// convert structured workload to unstructured
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(mpt.workload)
+	if err != nil {
+		return err
+	}
+	uv := reflect.ValueOf(u)
+
+	if err := mpt.setAt(mpt.mapping.Annotations, &mpt.Annotations, uv); err != nil {
+		return err
+	}
+	ci := 0
+	for i := range mpt.mapping.Containers {
+		cp := jsonpath.New("")
+		if err := cp.Parse(fmt.Sprintf("{%s}", mpt.mapping.Containers[i].Path)); err != nil {
+			return err
+		}
+		cr, err := cp.FindResults(u)
+		if err != nil {
+			// errors are expected if a path is not found
+			continue
+		}
+		for _, cv := range cr[0] {
```

Comment:
FindResults returns a slice whose length corresponding to the number of item in root node list for the template we sent into the Parse method. Because we are defining the root template there will only ever be a single item in this slice.